README.md
ABOUT this fork
---------------

this fork is used to "emulate" an chroot environment,
that means we keep the homedirectory on the normal file system,
put the user into an jail and mount his homedirectory into the jail
in that way we can use a normal webserver installation without chrooting everything

This is **EXPERIMENTAL**

ABOUT
-----

Jailkit is a set of utilities to limit user accounts to specific files 
using chroot() and or specific commands. Setting up a jail is a lot 
easier using these utilities.

See `man jailkit` for more detailed info on the different utilities 
in this package. If you have not yet installed jailkit you can see 
this man page using `man man/jailkit.8`.

REQUIREMENTS
------------

The most elementary tools are written in C and only need libc and 
libpthreads (available on almost any Linux system and most other posix
compliant systems).

The scripts are written in python, so they need python installed.

COPYRIGHT
---------

Jailkit is an open source project written by Olivier Sessink. It is 
released under a modified BSD licence.

EXAMPLE
-------

Suppose you need to create an account `test` that can do sftp and scp
only. You want it in a jail called `/home/sftproot` where it will have 
a homedirectory `/home/test`
```
#initialise the jail
mkdir /home/sftproot
jk_init -j /home/sftproot jk_lsh
jk_init -j /home/sftproot sftp
jk_init -j /home/sftproot scp
# create the account
adduser test
jk_jailuser -j /home/sftproot test
# edit the jk_lsh configfile in the jail (man jk_lsh)
# you can use every editor you want, I chose 'joe'
joe /home/sftproot/etc/jailkit/jk_lsh.ini
# now restart jk_socketd
killall jk_socketd
jk_socketd
# test the account
sftp test@localhost
# check the logs if everything is correct
tail /var/log/daemon.log /var/log/auth.log
```
NOTES ON VARIOUS APPLICATIONS

cvs - cvs needs /tmp/ for temporary files
procmail - needs /dev/null

COPYRIGHT
---------

Copyright (C) 2003, 2004, 2005, Olivier Sessink

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.


debian/changelog
jailkit (2.19-1) UNRELEASED; urgency=medium

  * somehow a bug URL was pasted into the ini file location in the sourcecode for jk_chrootsh in the 2.18 release. Fixed.

jailkit (2.18-1) UNRELEASED; urgency=medium

  * maintenance release, fix uid_t printing for very high uid numbers
  * minor improvements to jk_init.ini
  * add possibility to force --login in jk_chrootsh

 -- Olivier <olivier@marshall>  Wed, 18 Nov 2015 20:38:44 +0100

jailkit (2.17-1) precise; urgency=low

  * fix compiling with -lcap for some newer linux distributions
  * improve some argument checking and error messages
  * do not abort when called as -su
  * improve jk_socketd man page
  * make jk_cp honor option -j again

 -- Olivier <olivier@cort.fakenet>  Fri, 03 Jan 2014 23:43:26 +0100

jailkit (2.16-1) precise; urgency=low

  * fix compiling if -pthread and -lcap are both needed
  * fix bug in environment cleaning function
  * improve jk_init.ini for some 64bit linux distributions
  * improve error messages

 -- Olivier <olivier@cort.fakenet>  Thu, 18 Apr 2013 21:57:34 +0200

jailkit (2.15-1) lucid; urgency=low

  * use -pthreads if available instead of -lpthread
  * allow symlinks for /bin /lib and /sbin if they are moved into /usr/ (as fedora 17 has)

 -- Olivier Sessink <olivier@cort.fakenet>  Thu, 07 Jun 2012 23:20:26 +0200

jailkit (2.14-1) lucid; urgency=low

  * move jk_check to hashlib
  * fix for the prefix directory of jk_check
  * enable jk_chrootsh to be called su

 -- Olivier Sessink <olivier@cort.fakenet>  Thu, 28 Apr 2011 10:25:24 +0200

jailkit (2.13-1) lucid; urgency=low

  * fixes a regression in 2.12 - INIPREFIX was updated incorrectly in some cases 

 -- Olivier Sessink <olivier@cort.fakenet>  Sun, 10 Oct 2010 22:35:24 +0200

jailkit (2.12-1) lucid; urgency=low

  * jk_init and jk_cp can now resolve binaries using PATH

 -- Olivier Sessink <olivier@cort.fakenet>  Sun, 12 Sep 2010 21:30:23 +0200

jailkit (2.11-1) hardy; urgency=low

  * mostly documentation updates
  * minor Solaris specific improvements
  * fixed possible jk_lsh crash

 -- Olivier <olivier@cort.fakenet>  Sun, 07 Feb 2010 17:23:16 +0100

jailkit (2.10-1) hardy; urgency=low

  * 2.9 fix was incompatible with jk_jailuser, fixed now
  * compiler warning fixes

 -- Olivier <olivier@cort.fakenet>  Thu, 22 Oct 2009 13:31:18 +0200

jailkit (2.9-1) hardy; urgency=low

  * fix symlink handling, if a symlink inside the jail points to the real system jk_init and jk_cp could write files into the real system

 -- Olivier <olivier@cort.fakenet>  Wed, 14 Oct 2009 23:07:33 +0200

jailkit (2.8-1) hardy; urgency=low

  * support for capabilities to replace the setuid bit
  * minor Solaris compatibility fixes

 -- Olivier <olivier@cort.fakenet>  Thu, 20 Aug 2009 12:51:05 +0200

jailkit (2.7-1) hardy; urgency=low

  * fix a regression in jk_chrootsh and jk_uchroot for interactive sessions

 -- Olivier <olivier@cort.fakenet>  Sun, 05 Apr 2009 20:20:38 +0200

jailkit (2.6-1) hardy; urgency=low

  * some minor code cleanups and improvements
  * some fixes for Solaris compatibility

 -- Olivier <olivier@cort.fakenet>  Wed, 01 Apr 2009 21:13:20 +0200

jailkit (2.5-1) unstable; urgency=low

  * fixed an issue in jk_cp and jk_init that symlinks to directories were sometimes copied as directories instead of symlinks
  * improved some documentation
  * added a -j option to all utilities to improve consistency

 -- Olivier Sessink <olivier@aria.fakenet>  Fri, 07 Dec 2007 13:53:52 +0100

jailkit (2.4-1) unstable; urgency=low

  * added a new jk_uchroot utility
  * fixed a rare crash in configuration file parsing 
  * added new jk_init configuration file options
  * added new options to jk_cp and jk_init to retain ownership when copying files into the jail
  * improved the documentation
  * improved several chroot security checks 
  * fixed some build issues on OpenBSD
  * improved error reporting

 -- Olivier Sessink <olivier@cort.fakenet>  Tue, 17 Jul 2007 20:32:41 +0200

jailkit (2.3-1) unstable; urgency=low

  * various small fixes for the new jk_update utility
  * added a -k option to jk_cp, jk_init and jk_update to use hardlinks instead of copying
  * various fixes to install jailkit in a different location

 -- Olivier Sessink <olivier@aria.fakenet>  Thu, 28 Dec 2006 13:51:00 +0100

jailkit (2.2-1) unstable; urgency=low

  * Jailkit 2.2 released with a new utility jk_update to keep jails up-to-date
    and various improvements on *BSD operating systems

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Fri, 8 Dec 2006 00:00:00 +0200

jailkit (2.1-2) unstable; urgency=low

  * tarballs are uploaded again, there was a Makefile issue

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Fri, 8 Sep 2006 00:00:00

jailkit (2.1-1) unstable; urgency=low

  * improved logging
  * allowed sub-second intervals for jk_socketd
  * improved various checks in jk_jailuser
  * improved handling of recursive directories in jk_cp
  * new utility jk_list to list all jailed processes

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Thu, 7 Sep 2006 00:00:00

jailkit (2.0-1) unstable; urgency=low

  * fixes to allow jk_chrootsh to function on systems that use a name service
    caching daemon, or ldap authentication and others
  * some options for jk_chrootsh to relax some strict checks
  * replaced jk_addjailuser with jk_jailuser, which should work on more
    platforms
  * added the DEFAULT section to jk_lsh.ini and jk_chrootsh.ini to allow

src/jk_lib.h
/*
Copyright (c) 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2013 Olivier Sessink
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
  * Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
  * Redistributions in binary form must reproduce the above
    copyright notice, this list of conditions and the following
    disclaimer in the documentation and/or other materials provided
    with the distribution.
  * The names of its contributors may not be used to endorse or
    promote products derived from this software without specific
    prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.

*/

#ifndef __JK_LIB_H
#define __JK_LIB_H

#include "config.h"

#define JAIL_PREFIX "/chroot/"

#ifdef DEBUG
#define DEBUG_MSG printf
#else
#define DEBUG_MSG(args...)
 /**/
#endif

#ifdef DEBUG
#define DEBUG_LOG(args...) syslog(LOG_DEBUG, args)
#else
#define DEBUG_LOG(args...)
 /**/
#endif

#ifndef HAVE_MALLOC0
#define malloc0(size) memset(malloc(size),0,size)
#define HAVE_MALLOC0
#endif /* HAVE_MALLOC0 */


#define TESTPATH_NOREGPATH 1    // (0000 0001)
#define TESTPATH_GROUPW    2    // (0000 0010)
#define TESTPATH_OTHERW    4    // (0000 0100)
#define TESTPATH_SETUID    8    // (0000 1000)
#define TESTPATH_SETGID   16    // (0001 0000)
#define TESTPATH_OWNER    32    // (0010 0000)
#define TESTPATH_GROUP    64    // (0100 0000)

int file_exists(const char *path);
char *implode_array(char **arr, int arrlen, const char *delimiter);
char *ending_slash(const char *src);
int testsafepath(const char *path, int owner, int group);
int basicjailissafe(const char *path);
int dirs_equal(const char *dir1, const char *dir2);
int getjaildir(const char *oldhomedir, char **jaildir, char **newhomedir);
char *strip_string(char * string);
int count_char(const char *string, char lookfor);
char **explode_string(const char *string, char delimiter);
int count_array(char **arr);
void free_array(char **arr);
struct passwd *jk_fake_dir(struct passwd *pw);
void jk_mount (const char *jaildir, const char *home);
int jk_is_mounted (const char *path);
int jk_check_jail_owner (const char *jail, const char *user);
char *jk_extract_user (const char *path);
int jk_is_chrooted (const char *user);

#endif /* __JK_LIB_H */


src/jk_chrootsh.c
/*
 * the jailkit chroot() shell
 * this program does a safe chroot() and then executes the shell
 * that the user has within that new root (according to newroot/etc/passwd)
 *
 * I tried to merge some of the ideas from chrsh by Aaron D. Gifford,
 * start-stop-daemon from Marek Michalkiewicz and suexec by the Apache
 * group in this shell
 *

Copyright (c) 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2013, 2014, 2015 Olivier Sessink
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
  * Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
  * Redistributions in binary form must reproduce the above
    copyright notice, this list of conditions and the following
    disclaimer in the documentation and/or other materials provided
    with the distribution.
  * The names of its contributors may not be used to endorse or
    promote products derived from this software without specific
    prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <pwd.h>
#include <grp.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <errno.h>
#include <syslog.h>
#include <limits.h>
#include <signal.h>

#ifdef HAVE_SYS_CAPABILITY_H
#include <sys/capability.h>
#endif

/* #define DEBUG */

#ifdef DEBUG
#define DEBUG_MSG printf
#else
#define DEBUG_MSG(args...)
 /**/
#endif

#define PROGRAMNAME "jk_chrootsh"
#define CONFIGFILE INIPREFIX"/jk_chrootsh.ini"

#include "jk_lib.h"
#include "utils.h"
#include "iniparser.h"
#include "passwdparser.h"

/* doesn't compile on FreeBSD without this */
extern char **environ;

/*
typedef struct {
	char *key;
	char *value;
} Tsavedenv;

static Tsavedenv *savedenv_new(const char *key) {
	Tsavedenv *savedenv;
	char *val = getenv(key);
	if (!val) return NULL;
	savedenv = malloc(sizeof(Tsavedenv));
	savedenv->key = strdup(key);
	savedenv->value = strdup(val);
	return savedenv;
}

static void savedenv_restore(Tsavedenv *savedenv) {
	if (savedenv) {
		setenv(savedenv->key, savedenv->value, 1);
		DEBUG_MSG("restored %s=%s\n",savedenv->key, savedenv->value);
	}
}

static void savedenv_free(Tsavedenv *savedenv) {
	if (savedenv) {
		free(savedenv->key);
		free(savedenv->value);
		free(savedenv);
	}
}
*/

static int in_array(char **haystack, char * needle, int needlelen) {
	if (haystack && needle) {
		char **tmp = haystack;
		while (*tmp) {
			if (strlen(*tmp)==needlelen && strncmp(*tmp, needle, needlelen)==0) return 1;
			tmp++;
		}
	}
	return 0;
}

static void unset_environ_except(char **except) {
	char **tmp = environ;
	if (environ == NULL || *tmp==NULL || except==NULL || *except==NULL) {
		clearenv();
		return;
	}
	while (*tmp) {
		char* pos = strchr(*tmp, '=');
		if (pos == NULL) {
			/* invalid environment variable, are we being hacked? clear them all! */
			clearenv();
			return;
		}
		if (!in_array(except, *tmp, pos-*tmp)) {
			char *key = strndup(*tmp, pos-*tmp);
			unsetenv(key);
			free(key);
			tmp = environ;
			continue;
		}
		tmp++;
	}
}

static int have_capabilities(void) {
#ifdef HAVE_CAP_GET_PROC
	cap_t caps = cap_get_proc();
	if (caps) {
		cap_flag_value_t value_p;
		cap_get_flag(caps, CAP_SYS_CHROOT, CAP_EFFECTIVE,&value_p);
		cap_free(caps);
		return (value_p);
	}
#endif  /*HAVE_CAP_GET_PROC*/
	return 0;
}

void signal_handler(int signum) {
	syslog(LOG_ERR, "abort, received signal %s (%d)", strsignal(signum),signum);
	exit(666);
}

int main (int argc, char **argv) {
	unsigned int i;
	unsigned int use_capabilities=0;
	int ret;
	char **newargv;
	char *user=NULL, *tmp=NULL;

	struct passwd *pw=NULL;
	struct group *gr=NULL;
	unsigned long ngroups_max=NGROUPS_MAX;
	unsigned long ngroups=0;
	gid_t *gids;
	struct passwd *intpw=NULL; /* for internal_getpwuid() */
	char *jaildir=NULL, *newhome=NULL, *shell=NULL;
	Tiniparser *parser=NULL;
	char **envs=NULL;
	unsigned int relax_home_group_permissions=0;
	unsigned int relax_home_other_permissions=0;
	unsigned int relax_home_group=0;
	unsigned int relax_home_owner=0;
	unsigned int injail_login_shell=0;
	char *injail_shell=NULL;
	unsigned int skip_injail_passwd_check=0;

	DEBUG_MSG(PROGRAMNAME", started\n");
	/* open the log facility */
	openlog(PROGRAMNAME, LOG_PID, LOG_AUTH);

	/* attach signal handler */
	signal(SIGILL,signal_handler);
	signal(SIGSEGV,signal_handler);
	signal(SIGTERM,signal_handler);
	signal(SIGFPE,signal_handler);

	/* check if it PRORAMNAME that the user wants */
	tmp = strrchr(argv[0], '/');
	if (!tmp) {
		tmp = argv[0];
	} else {
		tmp++;
	}

	if (strcmp(tmp, PROGRAMNAME) != 0 && strcmp(tmp, "su")!= 0 && strcmp(tmp, "-su")!= 0 && strcmp(tmp, "-"PROGRAMNAME)!=0) {
		DEBUG_MSG("wrong name, tmp=%s, &tmp[1]=%s\n", tmp, &tmp[1]);
		syslog(LOG_ERR, "abort, "PROGRAMNAME" is called as %s", argv[0]);
		exit(1);
	}

	/* now test if we are setuid root (the effective user id must be 0, and the real user id > 0 */
	if (geteuid() != 0) {
		if (have_capabilities()) {
			use_capabilities=1;
		} else {
			syslog(LOG_ERR, "abort, effective user ID is not 0, possibly "PROGRAMNAME" is not setuid root");
			exit(11);
		}
	}
	if (getuid() == 0) {
		syslog(LOG_ERR, "abort, "PROGRAMNAME" is run by root, which does not make sense because user root can break out of a jail anyway");
		exit(12);
	}

	DEBUG_MSG("get user info\n");
	/* get user info based on the users name and not on the uid. this enables support
	for systems with multiple users with the same user id */
	tmp = getenv("USER");
	if (tmp && strlen(tmp)) {
		user = strdup(tmp);
	}
	if (user) {
		pw = getpwnam(user);
	} else {
		pw = getpwuid(getuid());
	}
	if (!pw) {
		syslog(LOG_ERR, "abort, failed to get user information for user ID %u: %s, check /etc/passwd", getuid(), strerror(errno));
		exit(13);
	}
	if (!pw->pw_name || strlen(pw->pw_name)==0) {
		syslog(LOG_ERR, "abort, got an empty username for user ID %u: %s, check /etc/passwd", getuid(), strerror(errno));
		exit(13);
	}
	if (user && strcmp(user,pw->pw_name)!=0) {
		syslog(LOG_ERR, "abort, asked for user %s, got user info for %s", user, pw->pw_name);
		exit(13);
	}
	if (pw->pw_uid != getuid()) {
		syslog(LOG_ERR, "abort, started by user ID %u, got user info %s with user ID %d,", getuid(), pw->pw_name, pw->pw_uid);
		exit(13);
	}
	DEBUG_MSG("got user %s\nget group info\n",pw->pw_name);
	gr = getgrgid(getgid());
	if (!gr) {
		syslog(LOG_ERR, "abort, failed to get group information for group ID %u: %s, check /etc/group", getgid(), strerror(errno));
		exit(13);
	}
	DEBUG_MSG("get additional groups\n");
	/* ngroups_max = sysconf(_SC_NGROUPS_MAX);*/
	gids = malloc(ngroups_max * sizeof(gid_t));
	ngroups = getgroups(ngroups_max,gids);
	if (ngroups == -1) {
		syslog(LOG_ERR, "abort, failed to get additional group information: %s, check /etc/group", strerror(errno));
		exit(13);
	}
#ifdef DEBUG
	printf("got additional groups ");
	for (i=0;i<ngroups;i++) {
		printf("%u, ",gids[i]);
	}
	printf("\n");
#endif

	/* make sure the jailkit config directory is owned root:root and not writable for others */
	if ( (testsafepath(INIPREFIX, 0, 0) &~TESTPATH_GROUPW) != 0 ) {
		syslog(LOG_ERR, "abort, jailkit configuration directory "INIPREFIX" is not safe; it should be owned 0:0 and not writable for others");
		exit(14);
	}
	parser = new_iniparser(CONFIGFILE);
	if (parser) {
		char *groupsec, *section=NULL, buffer[1024]; /* openbsd complains if this is <1024 */
		groupsec = strcat(strcpy(malloc0(strlen(gr->gr_name)+7), "group "), gr->gr_name);
		if (iniparser_has_section(parser, pw->pw_name)) {
			section = strdup(pw->pw_name);
		} else if (iniparser_has_section(parser, groupsec)) {
			section = groupsec;
		} else if (iniparser_has_section(parser, "DEFAULT")) {
			section = strdup("DEFAULT");
		}
		if (section != groupsec) free(groupsec);
		if (section) {
			unsigned int pos = iniparser_get_position(parser) - strlen(section) - 2;

			if (iniparser_get_string_at_position(parser, section, "env", pos, buffer, 1024) > 0) {
				envs = explode_string(buffer, ',');
			}
			relax_home_group_permissions = iniparser_get_int_at_position(parser, section, "relax_home_group_permissions", pos);
			relax_home_other_permissions = iniparser_get_int_at_position(parser, section, "relax_home_other_permissions", pos);
			relax_home_group = iniparser_get_int_at_position(parser, section, "relax_home_group", pos);
			relax_home_owner = iniparser_get_int_at_position(parser, section, "relax_home_owner", pos);
			if (iniparser_get_string_at_position(parser, section, "injail_shell", pos, buffer, 1024) > 0) {
				injail_shell = strdup(buffer);
			}
			if (injail_shell) {
				skip_injail_passwd_check = iniparser_get_int_at_position(parser, section, "skip_injail_passwd_check", pos);
			}
			injail_login_shell = iniparser_get_int_at_position(parser, section, "injail_login_shell", pos);
			DEBUG_MSG("section %s: relax_home_group_permissions=%d, relax_home_other_permissions=%d, relax_home_group=%d, injail_shell=%s, skip_injail_passwd_check=%d\n",
					section, relax_home_group_permissions, relax_home_other_permissions,
					relax_home_group, injail_shell, skip_injail_passwd_check);
			free(section);
		} else {
			DEBUG_MSG("no relevant section found in configfile\n");
		}
		iniparser_close(parser);
	} else {
		DEBUG_MSG("no configfile "CONFIGFILE" ??\n");
	}

	DEBUG_MSG("close filedescriptors\n");
	/* open file descriptors can be used to break out of a chroot, so we close all of them, except for stdin,stdout and stderr */
#ifdef OPEN_MAX
    i = OPEN_MAX;
#elif defined(NOFILE)
    i = NOFILE;
#else
    i = getdtablesize();
#endif
	while (--i > 2) {
		/*printf("closing file descriptor %d\n",i);*/
		while (close(i) != 0 && errno == EINTR);
	}
	/* now make sure file descriptors 0 1 and 2 are valid before we (or a child) starts writing to it */
	while (1) {
		int fd;
		fd = open("/dev/null", O_RDWR);
		if (fd < 0)
			exit(10);
		if (fd > 2) {
			close(fd);
			break;
		} else {
			DEBUG_MSG("re-opening file descriptor %d\n",fd);
		}
	}

	/* now we clear the environment, except for values allowed in /etc/jailkit/jk_chrootsh.ini */
	unset_environ_except(envs);
	if (envs) {
		free_array(envs);
	}

	if (pw->pw_gid != getgid()) {
		syslog(LOG_ERR, "abort, the group ID from /etc/passwd (%u) does not match the group ID we run with (%u)", pw->pw_gid, getgid());
		exit(15);
	}
	if (!pw->pw_dir || strlen(pw->pw_dir) ==0) {
		syslog(LOG_ERR, "abort, got an empty home directory for user %s (%u)", pw->pw_name, getuid());
		exit(16);
	}
	
	if (jk_is_chrooted(pw->pw_name) == 1) {
		/* modify user dir for jail */
		pw = jk_fake_dir(pw);
	}
	
	if (strstr(pw->pw_dir, "/./") == NULL) {
		syslog(LOG_ERR, "abort, homedir '%s' for user %s (%u) does not contain the jail separator <jail>/./<home>", pw->pw_dir, pw->pw_name, getuid());
		exit(17);
	}
	DEBUG_MSG("get jaildir\n");
	if (!getjaildir(pw->pw_dir, &jaildir, &newhome)) {
		syslog(LOG_ERR, "abort, failed to read the jail and the home from %s for user %s (%u)",pw->pw_dir, pw->pw_name, getuid());
		exit(17);
	}
	DEBUG_MSG("dir=%s,jaildir=%s,newhome=%s\n",pw->pw_dir, jaildir, newhome);
	DEBUG_MSG("get chdir()\n");
	if (chdir(jaildir) != 0) {
		syslog(LOG_ERR, "abort, chdir(%s) failed: %s, check the permissions for %s",jaildir,strerror(errno),jaildir);
		exit(19);
	} else {
		char test[1024];
		/* test if it really succeeded */
		if (getcwd(test, 1024)==NULL || !dirs_equal(jaildir, test)) {
			syslog(LOG_ERR, "abort, the current dir is %s after chdir(%s), but it should be %s",test,jaildir,jaildir);
			exit(21);
		}
	}

	/* here do test the ownership of the jail and the homedir and such
	the function testsafepath doe exit itself on any failure */
	if (!basicjailissafe(jaildir)) {
		syslog(LOG_ERR, "abort, %s is not a safe jail, check ownership and permissions.", jaildir);
		exit(53);
	}
	ret = testsafepath(pw->pw_dir, getuid(), getgid());
	if ((ret & TESTPATH_NOREGPATH) ) {
		syslog(LOG_ERR, "abort, path %s is not a directory", pw->pw_dir);
		exit(53);
	}
	if (!relax_home_owner && (ret & TESTPATH_OWNER) ) {
		syslog(LOG_ERR, "abort, path %s is not owned by %u", pw->pw_dir,getuid());
		exit(53);
	}
	if (!relax_home_group && (ret & TESTPATH_GROUP)) {
		syslog(LOG_ERR, "abort, path %s does not have group owner %u, set option 'relax_home_group' to relax this check", pw->pw_dir,getgid());
		exit(53);
	}
	if (!relax_home_group_permissions && (ret & TESTPATH_GROUPW)) {
		syslog(LOG_ERR, "abort, path %s is group writable, set option 'relax_home_group_permissions' to relax this check", pw->pw_dir);
		exit(53);
	}
	if (!relax_home_other_permissions && (ret & TESTPATH_OTHERW)) {
		syslog(LOG_ERR, "abort, path %s is writable for other, set option 'relax_home_other_permissions' to relax this check", pw->pw_dir);
		exit(53);
	}
	/* do a final log message */
	tmp = implode_array(&argv[1], argc-1, " ");
	syslog(LOG_INFO, "now entering jail %s for user %s (%u) with arguments %s", jaildir, pw->pw_name, getuid(), tmp);
	free(tmp);
	
	if (jk_is_chrooted(pw->pw_name) == 1) {
		// mount home dir into jail
		jk_mount(jaildir, newhome);
	}
	
	DEBUG_MSG("chroot()\n");
	/* do the chroot() call */
	if (chroot(jaildir)) {
		syslog(LOG_ERR, "abort, chroot(%s) failed: %s, check the permissions for %s", jaildir, strerror(errno), jaildir);
		exit(33);
	}

	if (use_capabilities) {
#ifdef HAVE_CAP_GET_PROC
		cap_t caps;
		cap_value_t capv[1];
		/* drop chroot capability, should we drop all other capabilities that may be used to escape from the jail too ?  */
		if ((caps = cap_get_proc()) == NULL) {
			syslog(LOG_ERR, "abort, failed to retrieve current capabilities: %s", strerror(errno));
			exit(101);
		}
		capv[0] = CAP_SYS_CHROOT;
		/* other capabilities that should/could be dropped:
		CAP_SETPCAP, CAP_SYS_MODULE, CAP_SYS_RAWIO, CAP_SYS_PTRACE, CAP_SYS_ADMIN */
		if (cap_set_flag(caps, CAP_PERMITTED, 1, capv, CAP_CLEAR)) {
			syslog(LOG_ERR, "abort, failed to set PERMITTED capabilities: %s", strerror(errno));
			exit(102);
		}
		if (cap_set_flag(caps, CAP_EFFECTIVE, 1, capv, CAP_CLEAR)) {
			syslog(LOG_ERR, "abort, failed to set effective capabilities: %s", strerror(errno));
			exit(103);
		}
		if (cap_set_flag(caps, CAP_INHERITABLE, 1, capv, CAP_CLEAR)) {
			syslog(LOG_ERR, "abort, failed to set INHERITABLE capabilities: %s", strerror(errno));
			exit(104);
		}
		if (cap_set_proc(caps)) {
			syslog(LOG_ERR, "abort, failed to apply new capabilities: %s", strerror(errno));
			exit(105);
		}
#else
		/* we should never get here */
		exit(333);
#endif
	} else {
		/* drop all privileges, it seems that we first have to setgid(),
			then we have to call initgroups(),
			then we call setuid() */
		if (setgid(getgid())) {
			syslog(LOG_ERR, "abort, failed to set effective group ID %u: %s", getgid(), strerror(errno));
			exit(34);
		}
		if (setgroups(ngroups, gids)==-1) {
			syslog(LOG_ERR, "abort, failed to set additional groups: %s", strerror(errno));
			exit(35);
		}
		free(gids);
	/*	if (initgroups(pw->pw_name, getgid())) {
			syslog(LOG_ERR, "abort, failed to init groups for user %s (%d), check %s/etc/group", pw->pw_name,getuid(),jaildir);
			exit(35);
		}*/
		if (setuid(getuid())) {
			syslog(LOG_ERR, "abort, failed to set effective user ID %u: %s", getuid(), strerror(errno));
			exit(36);
		}
	}
	/* test for user and group info, is it the same? checks username, groupname and home */
	if (!skip_injail_passwd_check){
		char *oldpw_name,*oldgr_name;
		oldpw_name = strdup(pw->pw_name);
		oldgr_name = strdup(gr->gr_name);

		if (user) {
			pw = getpwnam(user);
		} else {
			pw = getpwuid(getuid());
		}
		if (!pw) {
			syslog(LOG_ERR, "abort, failed to get user information in the jail for user ID %u: %s, check %s/etc/passwd",getuid(),strerror(errno),jaildir);
			exit(35);
		}
		if (pw->pw_uid != getuid()) {
			syslog(LOG_ERR, "abort, got user information in the jail for user ID %u instead of user ID %u, check %s/etc/passwd",pw->pw_uid,getuid(),jaildir);
			exit(35);
		}
		DEBUG_MSG("got %s as pw_dir\n",pw->pw_dir);
		gr = getgrgid(getgid());
		if (!gr) {
			syslog(LOG_ERR, "abort, failed to get group information in the jail for group ID %u: %s, check %s/etc/group",getgid(),strerror(errno),jaildir);
			exit(35);
		}
		if (strcmp(pw->pw_name, oldpw_name)!=0) {
			syslog(LOG_ERR, "abort, username %s differs from jail username %s for user ID %u, check /etc/passwd and %s/etc/passwd", oldpw_name, pw->pw_name, getuid(), jaildir);
			exit(37);
		}
		if (strcmp(gr->gr_name, oldgr_name)!=0) {
			syslog(LOG_ERR, "abort, groupname %s differs from jail groupname %s for group ID %u, check /etc/passwd and %s/etc/passwd", oldgr_name, gr->gr_name, getgid(), jaildir);
			exit(37);
		}
		if (strcmp(pw->pw_dir, newhome)!=0) {
			DEBUG_MSG("%s!=%s\n",pw->pw_dir, newhome);
			/* if these are different, it could be that getpwuid() gets the real user
			info (from for example ldap or nscd), and not the info inside the jail, lets
			test that, and if true, we should use the	shell from the internal function as well*/
			intpw = internal_getpwuid("/etc/passwd", getuid());
			if (!intpw) {
				DEBUG_MSG("%s!=%s\n",intpw->pw_dir, newhome);
				syslog(LOG_ERR, "abort, failed to find user %u in %s/etc/passwd", getuid(), jaildir);
				exit(39);
			}
			if (!dirs_equal(intpw->pw_dir, newhome)) {
				DEBUG_MSG("%s!=%s\n",intpw->pw_dir, newhome);
				syslog(LOG_ERR, "abort, home directory %s differs from jail home directory %s for user %s (%u), check /etc/passwd and %s/etc/passwd", newhome, pw->pw_dir, pw->pw_name, getuid(), jaildir);
				exit(39);
			}
		}
		free(oldpw_name);
		free(oldgr_name);
	}
	if (injail_shell) {
		shell = injail_shell;
	} else if (intpw) {
		shell = intpw->pw_shell;
	} else {
		shell = pw->pw_shell;
	}
	
	/* test the shell in the jail, it is not allowed to be setuid() root */
	testsafepath(shell,0,0);

	/* prepare the new environment */
	setenv("HOME",newhome,1);
	setenv("USER",pw->pw_name,1);
	setenv("USERNAME",pw->pw_name,1);
	setenv("SHELL",shell,1);
	if (chdir(newhome) != 0) {
		syslog(LOG_ERR, "abort, chdir(%s) failed inside the jail %s: %s, check the permissions for %s/%s",newhome,jaildir,strerror(errno),jaildir,newhome);
		exit(41);
	}
	
	/* cleanup before execution */
	free(newhome);

	/* now execute the jailed shell */
	if (injail_login_shell) {
		argc = 2;
		newargv = malloc0((argc+1)*sizeof(char *));
		newargv[0] = shell;
		newargv[1] = "--login";
	} else {
		newargv = malloc0((argc+1)*sizeof(char *));
		newargv[0] = shell;
		for (i=1;i<argc;i++) {
			newargv[i] = argv[i];
		}
	}
	execv(shell, newargv);
	DEBUG_MSG(strerror(errno));
	syslog(LOG_ERR, "ERROR: failed to execute shell %s for user %s (%u), check the permissions and libraries of %s/%s",shell,pw->pw_name,getuid(),jaildir,shell);

	free(jaildir);
	exit(111);
}
